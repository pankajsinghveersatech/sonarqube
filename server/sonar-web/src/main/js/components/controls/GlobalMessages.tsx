/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { keyframes } from '@emotion/core';
import styled from '@emotion/styled';
import * as React from 'react';
import { colors, sizes, zIndexes } from '../../app/theme';
import { cutLongWords } from '../../helpers/path';
import { ClearButton } from './buttons';

interface IMessage {
  id: string;
  level: 'ERROR' | 'SUCCESS';
  message: string;
}

export interface GlobalMessagesProps {
  closeGlobalMessage: (id: string) => void;
  messages: IMessage[];
}

export default function GlobalMessages({ closeGlobalMessage, messages }: GlobalMessagesProps) {
  if (messages.length === 0) {
    return null;
  }

  return (
    <MessagesContainer>
      {messages.map(message => (
        <GlobalMessage closeGlobalMessage={closeGlobalMessage} key={message.id} message={message} />
      ))}
    </MessagesContainer>
  );
}

const MessagesContainer = styled.div`
  position: fixed;
  z-index: ${zIndexes.processContainerZIndex};
  top: 0;
  left: 50%;
  width: 350px;
  margin-left: -175px;
`;

export class GlobalMessage extends React.PureComponent<{
  closeGlobalMessage: (id: string) => void;
  message: IMessage;
}> {
  handleClose = () => {
    this.props.closeGlobalMessage(this.props.message.id);
  };

  render() {
    const { message } = this.props;
    return (
      <Message
        data-test={`global-message__${message.level}`}
        level={message.level}
        role={message.level === 'SUCCESS' ? 'status' : 'alert'}>
        {cutLongWords(message.message)}
        <CloseButton
          className="button-small"
          color="#fff"
          level={message.level}
          onClick={this.handleClose}
        />
      </Message>
    );
  }
}

const appearAnim = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

const Message = styled.div<Pick<IMessage, 'level'>>`
  position: relative;
  padding: 0 30px 0 10px;
  line-height: ${sizes.controlHeight};
  border-radius: 0 0 3px 3px;
  box-sizing: border-box;
  color: #ffffff;
  background-color: ${({ level }) => (level === 'SUCCESS' ? colors.green : colors.red)};
  text-align: center;
  opacity: 0;
  animation: ${appearAnim} 0.2s ease forwards;

  & + & {
    margin-top: calc(${sizes.gridSize} / 2);
    border-radius: 3px;
  }
`;

const CloseButton = styled(ClearButton)<Pick<IMessage, 'level'>>`
  position: absolute;
  top: calc(${sizes.gridSize} / 4);
  right: calc(${sizes.gridSize} / 4);

  &:hover svg,
  &:focus svg {
    color: ${({ level }) => (level === 'SUCCESS' ? colors.green : colors.red)};
  }
`;
