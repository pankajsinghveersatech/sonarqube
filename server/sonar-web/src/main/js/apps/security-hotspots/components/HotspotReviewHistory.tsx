/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import { Button, ButtonLink, DeleteButton, EditButton } from '../../../components/controls/buttons';
import Dropdown, { DropdownOverlay } from '../../../components/controls/Dropdown';
import Toggler from '../../../components/controls/Toggler';
import DateTimeFormatter from '../../../components/intl/DateTimeFormatter';
import IssueChangelogDiff from '../../../components/issue/components/IssueChangelogDiff';
import Avatar from '../../../components/ui/Avatar';
import { PopupPlacement } from '../../../components/ui/popups';
import { translate, translateWithParameters } from '../../../helpers/l10n';
import { sanitizeString } from '../../../helpers/sanitize';
import { Hotspot, ReviewHistoryType } from '../../../types/security-hotspots';
import { getHotspotReviewHistory } from '../utils';
import HotspotCommentPopup from './HotspotCommentPopup';

export interface HotspotReviewHistoryProps {
  hotspot: Hotspot;
  onDeleteComment: (key: string) => void;
  onEditComment: (key: string, comment: string) => void;
  onShowFullHistory: () => void;
  showFullHistory: boolean;
}

export const MAX_RECENT_ACTIVITY = 5;

export default function HotspotReviewHistory(props: HotspotReviewHistoryProps) {
  const { hotspot, showFullHistory } = props;
  const fullReviewHistory = getHotspotReviewHistory(hotspot);
  const [editedCommentKey, setEditedCommentKey] = React.useState('');

  const reviewHistory = showFullHistory
    ? fullReviewHistory
    : fullReviewHistory.slice(0, MAX_RECENT_ACTIVITY);

  return (
    <>
      <ul>
        {reviewHistory.map((historyElt, historyIndex) => {
          const { user, type, diffs, date, html, key, updatable, markdown } = historyElt;
          return (
            <li
              className={classNames('padded-top padded-bottom', {
                'bordered-top': historyIndex > 0
              })}
              key={historyIndex}>
              <div className="display-flex-center">
                {user.name && (
                  <>
                    <Avatar
                      className="little-spacer-right"
                      hash={user.avatar}
                      name={user.name}
                      size={20}
                    />
                    <strong>
                      {user.active
                        ? user.name
                        : translateWithParameters('user.x_deleted', user.name)}
                    </strong>
                    {type === ReviewHistoryType.Creation && (
                      <span className="little-spacer-left">
                        {translate('hotspots.review_history.created')}
                      </span>
                    )}
                    {type === ReviewHistoryType.Comment && (
                      <span className="little-spacer-left">
                        {translate('hotspots.review_history.comment_added')}
                      </span>
                    )}
                    <span className="little-spacer-left little-spacer-right">-</span>
                  </>
                )}
                <DateTimeFormatter date={date} />
              </div>

              {type === ReviewHistoryType.Diff && diffs && (
                <div className="spacer-top">
                  {diffs.map((diff, diffIndex) => (
                    <IssueChangelogDiff diff={diff} key={diffIndex} />
                  ))}
                </div>
              )}

              {type === ReviewHistoryType.Comment && key && html && markdown && (
                <div className="spacer-top display-flex-space-between">
                  <div
                    className="markdown"
                    // eslint-disable-next-line react/no-danger
                    dangerouslySetInnerHTML={{ __html: sanitizeString(html) }}
                  />
                  {updatable && (
                    <div>
                      <div className="dropdown">
                        <Toggler
                          onRequestClose={() => {
                            setEditedCommentKey('');
                          }}
                          open={key === editedCommentKey}
                          overlay={
                            <DropdownOverlay placement={PopupPlacement.BottomRight}>
                              <HotspotCommentPopup
                                markdownComment={markdown}
                                onCancelEdit={() => setEditedCommentKey('')}
                                onCommentEditSubmit={comment => {
                                  setEditedCommentKey('');
                                  props.onEditComment(key, comment);
                                }}
                              />
                            </DropdownOverlay>
                          }>
                          <EditButton
                            className="button-small"
                            onClick={() => setEditedCommentKey(key)}
                          />
                        </Toggler>
                      </div>
                      <Dropdown
                        onOpen={() => setEditedCommentKey('')}
                        overlay={
                          <div className="padded abs-width-150">
                            <p>{translate('issue.comment.delete_confirm_message')}</p>
                            <Button
                              className="button-red big-spacer-top pull-right"
                              onClick={() => props.onDeleteComment(key)}>
                              {translate('delete')}
                            </Button>
                          </div>
                        }
                        overlayPlacement={PopupPlacement.BottomRight}>
                        <DeleteButton className="button-small" />
                      </Dropdown>
                    </div>
                  )}
                </div>
              )}
            </li>
          );
        })}
      </ul>
      {!showFullHistory && fullReviewHistory.length > MAX_RECENT_ACTIVITY && (
        <ButtonLink className="spacer-top" onClick={props.onShowFullHistory}>
          {translate('show_all')}
        </ButtonLink>
      )}
    </>
  );
}
