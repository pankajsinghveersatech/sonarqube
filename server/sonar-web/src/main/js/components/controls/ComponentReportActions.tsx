/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import {
  getReportStatus,
  subscribeToEmailReport,
  unsubscribeFromEmailReport
} from '../../api/component-report';
import addGlobalSuccessMessage from '../../app/utils/addGlobalSuccessMessage';
import { translate, translateWithParameters } from '../../helpers/l10n';
import { isLoggedIn } from '../../helpers/users';
import { Store } from '../../store/rootReducer';
import { Branch } from '../../types/branch-like';
import { ComponentQualifier } from '../../types/component';
import { ComponentReportStatus } from '../../types/component-report';
import { withCurrentUser } from '../hoc/withCurrentUser';
import ComponentReportActionsRenderer from './ComponentReportActionsRenderer';

interface Props {
  appState: Pick<T.AppState, 'qualifiers'>;
  component: T.Component;
  branch?: Branch;
  currentUser: T.CurrentUser;
}

interface State {
  loadingStatus?: boolean;
  status?: ComponentReportStatus;
}

export class ComponentReportActions extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = {};

  componentDidMount() {
    this.mounted = true;
    const governanceEnabled = this.props.appState.qualifiers.includes(ComponentQualifier.Portfolio);
    if (governanceEnabled) {
      this.loadReportStatus();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  loadReportStatus = async () => {
    const { component, branch } = this.props;

    const status = await getReportStatus(component.key, branch?.name).catch(() => undefined);

    if (this.mounted) {
      this.setState({ status, loadingStatus: false });
    }
  };

  handleSubscription = (subscribed: boolean) => {
    const { component } = this.props;
    const { status } = this.state;

    const translationKey = subscribed
      ? 'component_report.subscribe_x_success'
      : 'component_report.unsubscribe_x_success';
    const frequencyTranslation = translate(
      'report.frequency',
      status?.componentFrequency || status?.globalFrequency || ''
    ).toLowerCase();
    const qualifierTranslation = translate('qualifier', component.qualifier).toLowerCase();

    addGlobalSuccessMessage(
      translateWithParameters(translationKey, frequencyTranslation, qualifierTranslation)
    );

    this.loadReportStatus();
  };

  handleSubscribe = async () => {
    const { component, branch } = this.props;

    await subscribeToEmailReport(component.key, branch?.name);

    this.handleSubscription(true);
  };

  handleUnsubscribe = async () => {
    const { component, branch } = this.props;

    await unsubscribeFromEmailReport(component.key, branch?.name);

    this.handleSubscription(false);
  };

  render() {
    const { currentUser, component, branch } = this.props;
    const { status, loadingStatus } = this.state;

    if (loadingStatus || !status || (branch && !branch.excludedFromPurge)) {
      return null;
    }

    const currentUserHasEmail = isLoggedIn(currentUser) && !!currentUser.email;

    return (
      <ComponentReportActionsRenderer
        branch={branch}
        component={component}
        frequency={status.componentFrequency || status.globalFrequency}
        subscribed={status.subscribed}
        canSubscribe={status.canSubscribe}
        currentUserHasEmail={currentUserHasEmail}
        handleSubscription={this.handleSubscribe}
        handleUnsubscription={this.handleUnsubscribe}
      />
    );
  }
}

const mapStateToProps = (state: Store) => ({
  appState: state.appState
});

export default withCurrentUser(connect(mapStateToProps)(ComponentReportActions));
