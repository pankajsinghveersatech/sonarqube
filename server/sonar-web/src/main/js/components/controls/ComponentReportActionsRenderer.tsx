/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import * as React from 'react';
import { getReportUrl } from '../../api/component-report';
import { Button } from '../../components/controls/buttons';
import Dropdown from '../../components/controls/Dropdown';
import DropdownIcon from '../../components/icons/DropdownIcon';
import { translate, translateWithParameters } from '../../helpers/l10n';
import { Branch } from '../../types/branch-like';

export interface ComponentReportActionsRendererProps {
  component: T.Component;
  branch?: Branch;
  frequency: string;
  subscribed: boolean;
  canSubscribe: boolean;
  currentUserHasEmail: boolean;
  handleSubscription: () => void;
  handleUnsubscription: () => void;
}

export default function ComponentReportActionsRenderer(props: ComponentReportActionsRendererProps) {
  const { branch, component, frequency, subscribed, canSubscribe, currentUserHasEmail } = props;

  const renderDownloadButton = (simple = false) => {
    return (
      <a
        download={[component.name, branch?.name, 'PDF Report.pdf'].filter(s => !!s).join(' - ')}
        href={getReportUrl(component.key, branch?.name)}
        target="_blank"
        data-test="overview__download-pdf-report-button"
        rel="noopener noreferrer">
        {simple
          ? translate('download_verb')
          : translateWithParameters(
              'component_report.download',
              translate('qualifier', component.qualifier).toLowerCase()
            )}
      </a>
    );
  };

  const renderSubscriptionButton = () => {
    if (!currentUserHasEmail) {
      return (
        <span className="text-muted-2">{translate('component_report.no_email_to_subscribe')}</span>
      );
    }

    const translationKey = subscribed
      ? 'component_report.unsubscribe_x'
      : 'component_report.subscribe_x';
    const onClickHandler = subscribed ? props.handleUnsubscription : props.handleSubscription;
    const frequencyTranslation = translate('report.frequency', frequency).toLowerCase();

    return (
      <a href="#" onClick={onClickHandler} data-test="overview__subscribe-to-report-button">
        {translateWithParameters(translationKey, frequencyTranslation)}
      </a>
    );
  };

  return canSubscribe ? (
    <Dropdown
      overlay={
        <ul className="menu">
          <li>{renderDownloadButton(true)}</li>
          <li>{renderSubscriptionButton()}</li>
        </ul>
      }>
      <Button className="dropdown-toggle">
        {translateWithParameters(
          'component_report.report',
          translate('qualifier', component.qualifier)
        )}
        <DropdownIcon className="spacer-left icon-half-transparent" />
      </Button>
    </Dropdown>
  ) : (
    renderDownloadButton()
  );
}
