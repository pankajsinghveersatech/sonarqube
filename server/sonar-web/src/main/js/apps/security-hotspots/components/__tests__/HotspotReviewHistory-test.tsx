/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { Button, EditButton } from '../../../../components/controls/buttons';
import Dropdown, { DropdownOverlay } from '../../../../components/controls/Dropdown';
import Toggler from '../../../../components/controls/Toggler';
import { mockIssueChangelog } from '../../../../helpers/mocks/issues';
import { mockHotspot, mockHotspotComment } from '../../../../helpers/mocks/security-hotspots';
import { mockUser } from '../../../../helpers/testMocks';
import HotspotCommentPopup from '../HotspotCommentPopup';
import HotspotReviewHistory, { HotspotReviewHistoryProps } from '../HotspotReviewHistory';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ showFullHistory: true })).toMatchSnapshot('show full list');
  expect(
    shallowRender({ showFullHistory: true })
      .find(Toggler)
      .props().overlay
  ).toMatchSnapshot('edit comment overlay');
  expect(
    shallowRender({ showFullHistory: true })
      .find(Dropdown)
      .props().overlay
  ).toMatchSnapshot('delete comment overlay');
});

it('should correctly handle comment updating', () => {
  return new Promise<void>((resolve, reject) => {
    const setEditedCommentKey = jest.fn();
    jest.spyOn(React, 'useState').mockImplementationOnce(() => ['', setEditedCommentKey]);

    const onEditComment = jest.fn();
    const wrapper = shallowRender({ onEditComment, showFullHistory: true });

    // Closing the Toggler sets the edited key back to an empty string.
    wrapper
      .find(Toggler)
      .at(0)
      .props()
      .onRequestClose();
    expect(setEditedCommentKey).toBeCalledWith('');

    const editOnClick = wrapper
      .find(EditButton)
      .at(0)
      .props().onClick;
    if (!editOnClick) {
      reject();
      return;
    }

    // Clicking on the EditButton correctly flags the comment for editing.
    editOnClick();
    expect(setEditedCommentKey).toHaveBeenLastCalledWith('comment-1');

    // Cancelling an edit sets the edited key back to an empty string
    const dropdownOverlay = shallow(
      wrapper
        .find(Toggler)
        .at(0)
        .props().overlay as React.ReactElement<DropdownOverlay>
    );
    dropdownOverlay
      .find(HotspotCommentPopup)
      .props()
      .onCancelEdit();
    expect(setEditedCommentKey).toHaveBeenLastCalledWith('');

    // Updating the comment sets the edited key back to an empty string, and calls the
    // prop to update the comment value.
    dropdownOverlay
      .find(HotspotCommentPopup)
      .props()
      .onCommentEditSubmit('comment');
    expect(onEditComment).toHaveBeenLastCalledWith('comment-1', 'comment');
    expect(setEditedCommentKey).toHaveBeenLastCalledWith('');
    expect(setEditedCommentKey).toHaveBeenCalledTimes(4);

    resolve();
  });
});

it('should correctly handle comment deleting', () => {
  return new Promise<void>((resolve, reject) => {
    const setEditedCommentKey = jest.fn();
    jest.spyOn(React, 'useState').mockImplementationOnce(() => ['', setEditedCommentKey]);

    const onDeleteComment = jest.fn();
    const wrapper = shallowRender({ onDeleteComment, showFullHistory: true });

    // Opening the deletion Dropdown sets the edited key back to an empty string.
    const dropdownOnOpen = wrapper
      .find(Dropdown)
      .at(0)
      .props().onOpen;
    if (!dropdownOnOpen) {
      reject();
      return;
    }
    dropdownOnOpen();
    expect(setEditedCommentKey).toHaveBeenLastCalledWith('');

    // Confirming deletion calls the prop to delete the comment.
    const dropdownOverlay = shallow(
      wrapper
        .find(Dropdown)
        .at(0)
        .props().overlay as React.ReactElement<HTMLDivElement>
    );
    const deleteButtonOnClick = dropdownOverlay.find(Button).props().onClick;
    if (!deleteButtonOnClick) {
      reject();
      return;
    }

    deleteButtonOnClick();
    expect(onDeleteComment).toBeCalledWith('comment-1');

    resolve();
  });
});

function shallowRender(props?: Partial<HotspotReviewHistoryProps>) {
  return shallow(
    <HotspotReviewHistory
      hotspot={mockHotspot({
        creationDate: '2018-09-01',
        changelog: [
          mockIssueChangelog(),
          mockIssueChangelog({
            creationDate: '2018-10-12'
          })
        ],
        comment: [
          mockHotspotComment({
            key: 'comment-1',
            updatable: true
          }),
          mockHotspotComment({ key: 'comment-2', user: mockUser({ name: undefined }) }),
          mockHotspotComment({ key: 'comment-3', user: mockUser({ active: false }) }),
          mockHotspotComment({ key: 'comment-4' }),
          mockHotspotComment({ key: 'comment-5' })
        ]
      })}
      onDeleteComment={jest.fn()}
      onEditComment={jest.fn()}
      onShowFullHistory={jest.fn()}
      showFullHistory={false}
      {...props}
    />
  );
}
