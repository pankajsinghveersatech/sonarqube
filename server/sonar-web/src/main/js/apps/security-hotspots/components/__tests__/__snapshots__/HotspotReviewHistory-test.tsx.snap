// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<Fragment>
  <ul>
    <li
      className="padded-top padded-bottom"
      key="0"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="Luke Skywalker"
          size={20}
        />
        <strong>
          Luke Skywalker
        </strong>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-10-12"
        />
      </div>
      <div
        className="spacer-top"
      >
        <IssueChangelogDiff
          diff={
            Object {
              "key": "assign",
              "newValue": "darth.vader",
              "oldValue": "luke.skywalker",
            }
          }
          key="0"
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="1"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="Luke Skywalker"
          size={20}
        />
        <strong>
          Luke Skywalker
        </strong>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-10-01"
        />
      </div>
      <div
        className="spacer-top"
      >
        <IssueChangelogDiff
          diff={
            Object {
              "key": "assign",
              "newValue": "darth.vader",
              "oldValue": "luke.skywalker",
            }
          }
          key="0"
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="2"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="3"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="4"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          user.x_deleted.John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
      </div>
    </li>
  </ul>
  <ButtonLink
    className="spacer-top"
    onClick={[MockFunction]}
  >
    show_all
  </ButtonLink>
</Fragment>
`;

exports[`should render correctly: delete comment overlay 1`] = `
<div
  className="padded abs-width-150"
>
  <p>
    issue.comment.delete_confirm_message
  </p>
  <Button
    className="button-red big-spacer-top pull-right"
    onClick={[Function]}
  >
    delete
  </Button>
</div>
`;

exports[`should render correctly: edit comment overlay 1`] = `
<DropdownOverlay
  placement="bottom-right"
>
  <HotspotCommentPopup
    markdownComment="*TEST*"
    onCancelEdit={[Function]}
    onCommentEditSubmit={[Function]}
  />
</DropdownOverlay>
`;

exports[`should render correctly: show full list 1`] = `
<Fragment>
  <ul>
    <li
      className="padded-top padded-bottom"
      key="0"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="Luke Skywalker"
          size={20}
        />
        <strong>
          Luke Skywalker
        </strong>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-10-12"
        />
      </div>
      <div
        className="spacer-top"
      >
        <IssueChangelogDiff
          diff={
            Object {
              "key": "assign",
              "newValue": "darth.vader",
              "oldValue": "luke.skywalker",
            }
          }
          key="0"
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="1"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="Luke Skywalker"
          size={20}
        />
        <strong>
          Luke Skywalker
        </strong>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-10-01"
        />
      </div>
      <div
        className="spacer-top"
      >
        <IssueChangelogDiff
          diff={
            Object {
              "key": "assign",
              "newValue": "darth.vader",
              "oldValue": "luke.skywalker",
            }
          }
          key="0"
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="2"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="3"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="4"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          user.x_deleted.John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="5"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="john.doe"
          size={20}
        />
        <strong>
          john.doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="6"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.comment_added
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-10"
        />
      </div>
      <div
        className="spacer-top display-flex-space-between"
      >
        <div
          className="markdown"
          dangerouslySetInnerHTML={
            Object {
              "__html": "<strong>TEST</strong>",
            }
          }
        />
        <div>
          <div
            className="dropdown"
          >
            <Toggler
              onRequestClose={[Function]}
              open={false}
              overlay={
                <DropdownOverlay
                  placement="bottom-right"
                >
                  <HotspotCommentPopup
                    markdownComment="*TEST*"
                    onCancelEdit={[Function]}
                    onCommentEditSubmit={[Function]}
                  />
                </DropdownOverlay>
              }
            >
              <EditButton
                className="button-small"
                onClick={[Function]}
              />
            </Toggler>
          </div>
          <Dropdown
            onOpen={[Function]}
            overlay={
              <div
                className="padded abs-width-150"
              >
                <p>
                  issue.comment.delete_confirm_message
                </p>
                <Button
                  className="button-red big-spacer-top pull-right"
                  onClick={[Function]}
                >
                  delete
                </Button>
              </div>
            }
            overlayPlacement="bottom-right"
          >
            <DeleteButton
              className="button-small"
            />
          </Dropdown>
        </div>
      </div>
    </li>
    <li
      className="padded-top padded-bottom bordered-top"
      key="7"
    >
      <div
        className="display-flex-center"
      >
        <Connect(Avatar)
          className="little-spacer-right"
          name="John Doe"
          size={20}
        />
        <strong>
          John Doe
        </strong>
        <span
          className="little-spacer-left"
        >
          hotspots.review_history.created
        </span>
        <span
          className="little-spacer-left little-spacer-right"
        >
          -
        </span>
        <DateTimeFormatter
          date="2018-09-01"
        />
      </div>
    </li>
  </ul>
</Fragment>
`;
